lib/checkautovariables.cpp

```diff
@@ -628,11 +584,22 @@ void CheckAutoVariables::checkVarLifetimeScope(const Token * start, const Token
             } else if (isDeadScope(val.tokvalue->variable()->nameToken(), tok->scope())) {
                 errorInvalidLifetime(tok, &val);
                 break;
-            } else if (tok->variable() && tok->variable()->declarationId() == tok->varId() &&
-                       !tok->variable()->isLocal() && !tok->variable()->isArgument() &&
-                       isInScope(val.tokvalue->variable()->nameToken(), tok->scope())) {
-                errorDanglngLifetime(tok, &val);
-                break;
+            } else if (isInScope(val.tokvalue->variable()->nameToken(), tok->scope())) {
+                const Variable * var = nullptr;
+                const Token * tok2 = tok;
+                if (Token::simpleMatch(tok->astParent(), "=")) {
+                    if (tok->astParent()->astOperand2() == tok) {
+                        var = getLHSVariable(tok->astParent());
+                        tok2 = tok->astParent()->astOperand1();
+                    }
+                }
+                else if (tok->variable() && tok->variable()->declarationId() == tok->varId()) {
```

Comment:
could you point out the situation when you see different `declarationId` and `varId`?

Reply:
Yea it happens on this case:

```cpp
void f() {
    struct b {
        uint32_t f[6];
    } d;
    uint32_t *a = d.f;
}
```

And this test will fail without the condition there.